### DISCUSSION TITLE
gRPC request body placeholder

### DISCUSSION THREAD

**Discussion by gAmUssA** [Kong/insomnia#3475] [Ideas]
Hi team,

I would love if Insomnia created request body placeholder when testing gRPC service instead of having empty request. 
This information is available in proto file.

Thanks!

**Comment by MatthewFin** [comment]
+1 for this feature. BloomRPC does this and it's super useful
![image](https://user-images.githubusercontent.com/17120263/124258563-d6483600-db2d-11eb-8dc3-f3d13c75c17c.png)

**Reply by gianlucaparadise** [reply]
+1

grpcui and Kreya also do it and it is a time saver

**Comment by vispariana** [comment]
It's really useful and all major competitors have already implemented this. There is also another discussion about this feature at #4874 with 5 up-votes.

However, there are two things to consider about this feature:
#### 1.  When to generate the stub
Postman brings up an empty editor with a **Generate Example Message** button underneath. Another option is to generate the stub when proto file is selected or server reflection is used.

#### 2. What the stub should look like
Postman prefills the fields like this:
- Numeral fields are filled with large random numbers (which has been useless and bothering to me since most numbers I used in my APIs were either zero or < 100 and I had to rewrite all of them)
- `string` fields are filled with random phrases from lorem ipsum
- `repeated` fields have three items inside

I suggest generating stubs on proto change (either by uploading new proto or using server reflection) to remove the need for a UI change
As for how the stub should look like, my personal suggestion is having one item for `repeated` fields and initializing all other fields with their zero value.

P.S) Just curious, if we choose to go on with the "generate example message" button approach, who would be doing (or confirming) the UI design?

**Comment by jackkav** [comment]
References for anyone who might like to implement this.

https://github.com/bloomrpc/bloomrpc/blob/a5c457147809fe807523bcb42939ae0fb227c5cd/app/behaviour/importProtos.ts#L186

https://github.com/bloomrpc/bloomrpc-mock/blob/master/src/automock.ts#L93

**Comment by jerroldgao** [comment]
This is super helpful! Since BloomRPC is deprecated, I wish Insomnia can bring this feature

**Comment by alvian-pratama** [comment]
note this should already be provided by grpc reflection, for example on grpcURL there's a way to [get the structure of grpc endpoint](https://github.com/grpc/grpc-go/blob/master/Documentation/server-reflection-tutorial.md#describe-services-and-methods) and [get the detail of each message](https://github.com/grpc/grpc-go/blob/master/Documentation/server-reflection-tutorial.md#inspect-message-types)

**Comment by Sargastico** [comment]
any updates?

**Comment by nbgraham** [comment]
I attempted to implement this here: https://github.com/Kong/insomnia/pull/6196

**Comment by jackkav** [comment]
Merged! Thanks again @nbgraham
